lib/logstash/inputs/unix.rb

```diff
@@ -140,23 +138,26 @@ def run(output_queue)
             raise
           end
         end
-      end # loop
+      end
     else
-      loop do
-        client_socket = UNIXSocket.new(@path)
-        client_socket.instance_eval { class << self; include ::LogStash::Util::SocketPeer end }
+      while !stop?
+        @client_socket = UNIXSocket.new(@path)
+        @client_socket.instance_eval { class << self; include ::LogStash::Util::SocketPeer end }
         @logger.debug("Opened connection", :client => @path)
-        handle_socket(client_socket, output_queue)
-      end # loop
+        handle_socket(@client_socket, output_queue)
+      end
     end
   end # def run
 
   public
-  def teardown
+  def stop
+    super
     if server?
       File.unlink(@path)
       @interrupted = true
       @thread.raise(Interrupted.new)
+    else
+      @client_socket.close
     end
-  end # def teardown
+  end
```

Comment:
This code cry for refactoring, I am trying to make sure we do the minimal changes for 2.0 and only introduced needed changes.

The reimplementation of the `#stop` method should respect the contract of the base class which follow thread safety, I notice that because the `@thread`  ivar is instantiated in the `#run` and also used in the stop, which will be called by two different threads. We need to make sure that we cross the memory barrier when accessing  them.